Add Beat CRD docs (#3362)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/beat.asciidoc

:page_id: beat
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Run Beats on ECK

This section describes how to configure and deploy Beats with ECK.

* <<{p}-beat-quickstart,Quickstart>>
* <<{p}-beat-configuration,Configuration>>
* <<{p}-beat-troubleshooting,Troubleshooting>>

[id="{p}-beat-quickstart"]
== Quickstart

. Apply the following specification to deploy Filebeat and collect the logs of all containers running in the Kubernetes cluster. ECK automatically configures the secured connection to an Elasticsearch cluster named `quickstart`, created in the link:k8s-quickstart.html[Elasticsearch quickstart].
+
[source,yaml,subs="attributes,+macros"]
----
cat $$<<$$EOF | kubectl apply -f -
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: filebeat-quickstart
spec:
  type: filebeat
  version: {version}
  elasticsearchRef:

>     name: elasticsearch-quickstart

Reviewer: This should be `quickstart` to align with the Elasticsearch Quickstart guide

  config:
    filebeat.inputs:
    - type: container
      paths:
      - /var/log/containers/*.log
  daemonSet:
    podTemplate:
      spec:
        dnsPolicy: ClusterFirstWithHostNet
        hostNetwork: true
        securityContext:
          runAsUser: 0
        containers:
        - name: filebeat
          volumeMounts:
          - name: varlogcontainers
            mountPath: /var/log/containers
          - name: varlogpods
            mountPath: /var/log/pods
          - name: varlibdockercontainers
            mountPath: /var/lib/docker/containers
        volumes:
        - name: varlogcontainers
          hostPath:
            path: /var/log/containers
        - name: varlogpods
          hostPath:
            path: /var/log/pods
        - name: varlibdockercontainers
          hostPath:
            path: /var/lib/docker/containers
EOF
----

> . Monitor Beats

Reviewer 2: I wonder if there's a better verb here than "monitor". "Inspect Beat status" maybe?

Author: This is following the other quickstarts we have, we've used Monitor in this section so far.

- `image` element in the specification must point to the image to be deployed
-  a role must exist in Elasticsearch and have the permissions required by the Beat. The role name must be `eck_beat_$type_role`, where `$type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`. See the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html[Elasticsearch documentation] for more details.

[id="{p}-beat-set-up-kibana-dashboards"]
=== Set up Kibana dashboards

ECK can instruct Beats to set up example dashboards packaged with the Beat. To enable this, set the `kibanaRef` element in the specification to point to ECK-managed Kibana deployment:

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  kibanaRef:
    name: quickstart
...
----

ECK will create a new user in Elasticsearch with a minimal set of appropriate roles and permissions that is needed for dashboard setup.

[id="{p}-beat-secrets-keystore-for-secure-settings"]
=== Secrets keystore for secure settings

> Beats offer a secret keystore for sensitive settings that need to be provided in the config, like for example passwords. This avoids storing them in the config directly.

Reviewer 2: ```suggestion
Beats offer a secret keystore for sensitive settings that need to be provided in the config, for example passwords. This avoids storing them in the config directly.
```

  config:
    filebeat:
      autodiscover:
        providers:
        - host: ${HOSTNAME}
          type: kubernetes
          hints:
            enabled: true
            default_config:
              type: container
              paths:
              - /var/log/containers/*${data.kubernetes.container.id}.log
  daemonSet:
    podTemplate:
      spec:
        serviceAccount: elastic-beat-filebeat-quickstart
        automountServiceAccountToken: true
...
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elastic-beat-filebeat-quickstart
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: elastic-beat-autodiscover-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elastic-beat-autodiscover
subjects:
- kind: ServiceAccount
  name: elastic-beat-filebeat-quickstart
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elastic-beat-autodiscover
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  - namespaces
  - events
  - pods
  verbs:
  - get
  - list
  - watch
----

[id="{p}-beat-deploying-beats-in-secured-clusters"]
=== Deploying Beats in secured clusters

> To deploy Beats in clusters with the Pod Security Policy admission controller enabled, or in OpenShift clusters, you have to grant additional permissions to the Service Account that the Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.

Reviewer 2: ```suggestion
To deploy Beats in clusters with the Pod Security Policy admission controller enabled, or in OpenShift clusters, you must grant additional permissions to the Service Account used by the Beat Pods. Those Service Accounts must be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.
```

Maybe spell out SCC the first time we use it here too

          - name: varlibdockercontainers
            mountPath: /var/lib/docker/containers
        volumes:
        - name: varlogcontainers
          hostPath:
            path: /var/log/containers
        - name: varlogpods
          hostPath:
            path: /var/log/pods
        - name: varlibdockercontainers
          hostPath:
            path: /var/lib/docker/containers
EOF
----

. Monitor Beats
+
Retrieve details about the Filebeat:
+
[source,sh]
----
kubectl get beat
----
+
[source,sh,subs="attributes"]
----
NAME                  HEALTH   AVAILABLE   EXPECTED   TYPE       VERSION   AGE
quickstart            green    3           3          filebeat   {version}     2m
----
+
List all the Pods belonging to a given Beat:
+
[source,sh]
----
kubectl get pods --selector='beat.k8s.elastic.co/name=quickstart-beat-filebeat'
----
+
[source,sh]
----
NAME                                      READY   STATUS    RESTARTS   AGE
quickstart-beat-filebeat-tkz65            1/1     Running   0          3m45s
quickstart-beat-filebeat-kx5jt            1/1     Running   0          3m45s
quickstart-beat-filebeat-nb6qh            1/1     Running   0          3m45s
----
+

> . Access logs for one of the Pods

Reviewer 2: ```suggestion
. Access logs for one of the Pods:
```

kind: Beat
metadata:
  name: heartbeat-quickstart
spec:
  type: heartbeat
  version: 7.8.0
  elasticsearchRef:
    name: quickstart
  configRef:
    secretName: heartbeat-config
  deployment:
    podTemplate:
      spec:
        dnsPolicy: ClusterFirstWithHostNet
        securityContext:
          runAsUser: 0
---
apiVersion: v1
kind: Secret
metadata:
  name: heartbeat-config
stringData:
  beat.yml: |-
    heartbeat.monitors:
    - type: tcp
      schedule: '@every 5s'
      hosts: ["quickstart-es-http.default.svc:9200"]
----

For more details about Beats configuration, see the link:https://www.elastic.co/guide/en/beats/libbeat/current/beats-reference.html[Beats documentation].

[id="{p}-beat-deploy-elastic-beat"]
=== Deploy a Beat

ECK supports the deployment of the following Beats:

- link:https://www.elastic.co/beats/filebeat[Filebeat]
- link:https://www.elastic.co/beats/metricbeat[Metricbeat]
- link:https://www.elastic.co/beats/heartbeat[Heartbeat]
- link:https://www.elastic.co/beats/auditbeat[Auditbeat]
- link:https://www.elastic.co/beats/packetbeat[Packetbeat]
- link:https://www.elastic.co/guide/en/beats/journalbeat/current/index.html[Journalbeat]

> You can specify the Beat to deploy and its version through `type` and `version` elements. ECK computes the correct path to the container image, so the `image` element doesn't have to be provided. ECK creates a new user in Elasticsearch with a minimal set of appropriate roles and permissions to enable the use of all Beats features.

Reviewer 2: ```suggestion
You can specify the Beat to deploy and its version through `type` and `version` elements. ECK computes the correct path to the container image, so the `image` element can be omitted. ECK creates a new user in Elasticsearch with a minimal set of appropriate roles and permissions to enable the use of all Beats features.
```

though nowhere in the docs do we even reference the `image` field before now so I wonder if this is even necessary here. Maybe it can just be mentioned as a clarification in the next community beat section?

[id="{p}-beat-deploy-community-beat"]
=== Deploy a Community Beat

ECK supports the deployment of any Community Beat. `type` and `version` specification elements have to be provided. In addition:

- `image` must point to the image to be deployed

> -  a role must exist in Elasticsearch and have the permissions required by the Beat. The role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.

Reviewer 2: ```suggestion
-  a role must exist in Elasticsearch and have the permissions required by the Beat. The role name must be `eck_beat_$type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.
```

Or maybe `TYPE`?

Reviewer 2: Maybe worth linking to the ECK role docs here too and/or providing an example?

Beats offer a secret keystore for sensitive settings that need to be provided in the config, like for example passwords. This avoids storing them in the config directly.

ECK exposes that mechanism with the `secureSettings` element in the specification. <<{p}-es-secure-settings,Similar to Elasticsearch>>, you can use Kubernetes Secrets to provide the settings in a secure way:

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  secureSettings:
  - secretName: agent-name-secret
  config:
    name: ${AGENT_NAME_VAR}
...
---
apiVersion: v1
kind: Secret
metadata:
  name: agent-name-secret
stringData:
  AGENT_NAME_VAR: id_007
----

See Beats link:https://www.elastic.co/guide/en/beats/filebeat/current/keystore.html[documentation] for more details.

[id="{p}-beat-set-beat-output"]
=== Set Beat output

If the `elasticsearchRef` element is specified, ECK populates the output section of the Beat config. ECK creates a user with appropriate roles and permissions and uses its credentials. If required, it also mounts the CA certificate in all Beat Pods, and recreates Pods when this certificate changes.

> Output can be set to any that's supported by a given Beat. To do that, remove the `elasticsearchRef` element from the specification and include an appropriate output configuration in the `config` or `configRef`.

Reviewer 2: ```suggestion
The `output` element can be set to any value that is supported by a given Beat. To use it, remove the `elasticsearchRef` element from the specification and include an appropriate `output` configuration in the `config` or `configRef` elements.
```

Author: I changed it a bit differently, I feel that referring to output part of Beat config in the same way as to CRD fields ("`output` element") might be confusing as there is no `output` in the CRD.

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  config:
    output.redis:
      hosts: ["my_redis.default.svc"]

>       password: "my_password"

Reviewer 2: I think we probably do not want to encourage putting passwords in the config directly and should have an example here using configRef (and maybe even removing the text about it being possible to place in the config)

Author: Should I just change it to a config that doesn't contain any credentials? I worry that using `configRef` with a Secret will be confusing and removing `config` will imply that setting custom output is only possible via `configRef`.

Reviewer 2: Well a custom output that requires credentials should not be in the `config` and should only be in the `configRef`. I'm good with whatever as long as we don't promote storing passwords in plaintext

Author: This can be true even when using `config` if user relies on env variables or Beats keystore. I've change it to kafka config that doesn't need to contain creds.

      key: "filebeat"
      db: 0
      timeout: 5
...
----

[id="{p}-beat-chose-the-deployment-model"]
=== Choose the deployment model

Depending on the use case, Beats may need to be deployed as a link:https://kubernetes.io/docs/concepts/workloads/controllers/deployment/[Deployment] or a link:https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/[DaemonSet]. Provide a `podTemplate` element under either the `deployment` or the `daemonSet` element in the specification to choose how a given Beat should be deployed.

[id="{p}-beat-role-based-access-control-for-beats"]
=== Role Based Access Control for Beats

> Some Beats features (such as link:https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover.html[autodiscover] or Kubernetes module link:https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-kubernetes-apiserver.html[metricsets]) require Beat Pods to interact with Kubernetes APIs. Specific permissions are needed to allow that. Standard Kubernetes link:https://kubernetes.io/docs/reference/access-authn-authz/rbac/[RBAC] rules apply. For example, to allow for autodiscover:

Reviewer 2: ```suggestion
Some Beats features (such as link:https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover.html[autodiscover] or Kubernetes module link:https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-kubernetes-apiserver.html[metricsets]) require that Beat Pods interact with Kubernetes APIs. Specific permissions are needed to allow this functionality. Standard Kubernetes link:https://kubernetes.io/docs/reference/access-authn-authz/rbac/[RBAC] rules apply. For example, to allow for autodiscover:
```

kind: ClusterRoleBinding
metadata:
  name: elastic-beat-autodiscover-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elastic-beat-autodiscover
subjects:
- kind: ServiceAccount
  name: elastic-beat-filebeat-quickstart
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elastic-beat-autodiscover
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  - namespaces
  - events
  - pods
  verbs:
  - get
  - list
  - watch
----

[id="{p}-beat-deploying-beats-in-secured-clusters"]
=== Deploying Beats in secured clusters

To deploy Beats in clusters with the Pod Security Policy admission controller enabled, or in OpenShift clusters, you have to grant additional permissions to the Service Account that the Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.

[id="{p}-beat-troubleshooting"]
== Troubleshooting

[id="{p}-beat-beat-pods-are-crashing-when-kibanaref-is-specified"]
=== Beat Pods are crashing when kibanaRef is specified

> When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Pod crashes. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The pod may cycle at most a few times during the first initialization. Afterwards, the Beat should run successfully.

Reviewer 2: ```suggestion
When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Pod crashes. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The Pod may resart a few times during the first initialization. Afterwards, the Beat should run successfully.
```

[id="{p}-beat-deploy-community-beat"]
=== Deploy a Community Beat

ECK supports the deployment of any Community Beat. `type` and `version` specification elements have to be provided. In addition:

- `image` must point to the image to be deployed
-  a role must exist in Elasticsearch and have the permissions required by the Beat. The role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.

[id="{p}-beat-set-up-kibana-dashboards"]
=== Set up Kibana dashboards

ECK can instruct Beats to set up example dashboards packaged with the Beat. To enable this, set the `kibanaRef` element in the specification to point to ECK-managed Kibana deployment:

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  kibanaRef:
    name: quickstart
...
----

ECK will create a new user in Elasticsearch with a minimal set of appropriate roles and permissions that is needed for dashboard setup.

[id="{p}-beat-secrets-keystore-for-secure-settings"]
=== Secrets keystore for secure settings

> 

Reviewer 3: nit: extra newline

> Beats offer a secret keystore for sensitive settings that need to be provided in the configuration, like for example passwords. This avoids storing them in the config directly.

Reviewer 3: `in the configuration` vs. `in the config` in the same sentence feels inconsistent

ECK exposes that mechanism with the `secureSettings` element in the specification. <<{p}-es-secure-settings,Similar to Elasticsearch>>, you can use Kubernetes Secrets to provide the settings in a secure way:

> 

Reviewer 3: nit: extra newline

    name: ${AGENT_NAME_VAR}
...
---
apiVersion: v1
kind: Secret
metadata:
  name: agent-name-secret
stringData:
  AGENT_NAME_VAR: id_007
----

See Beats link:https://www.elastic.co/guide/en/beats/filebeat/current/keystore.html[documentation] for more details.

[id="{p}-beat-set-beat-output"]
=== Set Beat output

If the `elasticsearchRef` element is specified, ECK populates the output section of the Beat config. ECK creates a user with appropriate roles and permissions and uses its credentials. If required, it also mounts the CA certificate in all Beat Pods, and recreates Pods when this certificate changes.

Output can be set to any that's supported by a given Beat. To do that, remove the `elasticsearchRef` element from the specification and include an appropriate output configuration in the `config` or `configRef`.

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  config:
    output.redis:
      hosts: ["my_redis.default.svc"]
      password: "my_password"
      key: "filebeat"
      db: 0
      timeout: 5
...
----

[id="{p}-beat-chose-the-deployment-model"]
=== Choose the deployment model

> Depending on the use case, Beats may need to be deployed as a Deployment or a DaemonSet. Provide a `podTemplate` element under either the `deployment` or the `daemonSet` element in the specification to choose how a given Beat should be deployed.

Reviewer 3: Maybe add a link on `Deployment` and `DaemonSet` to their k8s official docs.

:page_id: beat
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Run Beats on ECK

This section describes how to configure and deploy Beats with ECK.

* <<{p}-beat-quickstart,Quickstart>>
* <<{p}-beat-configuration,Configuration>>
* <<{p}-beat-troubleshooting,Troubleshooting>>

[id="{p}-beat-quickstart"]
== Quickstart

> . Apply the following specification to deploy Filebeat. ECK automatically configures the secured connection to an Elasticsearch cluster named `quickstart`, created in the link:k8s-quickstart.html[Elasticsearch quickstart].

Reviewer 3: I think it'd be worth describing a little bit what this quickstart is doing.

```
Apply the following specification to deploy Filebeat and collect the logs of all containers running in the Kubernetes cluster.
```

> +

Reviewer 3: I was thinking our quickstart would be more like this (less yaml):

```
Run the following specification to deploy Filebeat and Metricbeat on your Kubernetes cluster. Both are setup to collect logs from all Pod, metrics from all hosts, and ship data to an Elasticsearch cluster named `quickstart`, created in the link:k8s-quickstart.html[Elasticsearch quickstart].

kubectl apply -f github.com/elastic/cloud-on-k8s/link/to/k8s-observability-recipe.yaml
```

Author: It is fairly short for what it does (compared to non-eck approach), so I thought it might be good to show that. If we just have a link, then it's not immediately obvious whether it's just our resource or also bunch of other things. WDYT? We will also have a recipes section that will have links only.

Reviewer: I think I have changed my mind and agree with David that it is actually short enough to be included directly.

Reviewer 3: 👍 fine with me as well, it has the benefit of making people understand the CRD logic.
Let's make sure we include a link to the recipes (once we have them) in the quickstart so people can jump quickly to the interesting manifests.

[source,yaml,subs="attributes,+macros"]
----
cat $$<<$$EOF | kubectl apply -f -
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: filebeat-quickstart
spec:
  type: filebeat
  version: {version}
  elasticsearchRef:
    name: elasticsearch-quickstart
  config:
    filebeat.inputs:
    - type: container
      paths:
      - /var/log/containers/*.log
  daemonSet:
    podTemplate:
      spec:
        dnsPolicy: ClusterFirstWithHostNet
        hostNetwork: true
        securityContext:
          runAsUser: 0
        containers:
        - name: filebeat
          volumeMounts:
          - name: varlogcontainers
            mountPath: /var/log/containers
          - name: varlogpods
            mountPath: /var/log/pods
          - name: varlibdockercontainers
            mountPath: /var/lib/docker/containers
        volumes:
        - name: varlogcontainers
          hostPath:
            path: /var/log/containers
        - name: varlogpods
          hostPath:
            path: /var/log/pods
        - name: varlibdockercontainers
          hostPath:

>             path: /var/lib/docker/containers

Author: Yes. No autodiscover and not adding k8s metadata.

>             path: /var/lib/docker/containers

Reviewer 2: I think the k8s metadata is really useful, I wonder if we want to provide this as a quickstart but link to a more complicated example that I assume 99+% of users will want to use.

>             path: /var/lib/docker/containers

Author: I will add that together with recipes library (coming soon).

filebeat-quickstart-beat-filebeat-kx5jt   1/1     Running   0          3m45s
filebeat-quickstart-beat-filebeat-nb6qh   1/1     Running   0          3m45s
----
+
. Access logs for one of the Pods
+
[source,sh]
----
kubectl logs -f filebeat-quickstart-beat-filebeat-tkz65
----

. Access logs ingested by Filebeat
+
- follow the Elasticsearch deployment link:k8s-deploy-elasticsearch.html[guide] and run:
+
[source,sh]
----
curl -u "elastic:$PASSWORD" -k "https://localhost:9200/filebeat-*/_search"
----
+
- or follow the Kibana deployment link:k8s-deploy-kibana.html[guide], log in and go to *Kibana* > *Discover*.

. Inspect generated Filebeat configuration
+
[source,sh]
----
kubectl get secret filebeat-quickstart-beat-filebeat-config -o jsonpath='{.data.beat\.yml}' | base64 --decode
----
+
[source,sh]
----
filebeat:
  inputs:
  - paths:
    - /var/log/containers/*.log
    type: container
output:
  elasticsearch:
    hosts:
    - https://elasticsearch-quickstartquickstart-es-http.default.svc:9200
    password: <redacted>
    ssl:
      certificate_authorities:
      - /mnt/elastic-internal/elasticsearch-certs/ca.crt
    username: default-filebeat-quickstart-beat-user

> ----

Reviewer 3: Do we need that part? We don't show this sort of example for other CRDs.
I understand why we'd want it to understand and deal with config merging but since we don't setup a default config besides the es output it feels unnecessary.

Author: Since we know about issues with `key: null` I wanted to hint how to look at the final config that Beat uses. But maybe that should be moved to troubleshooting section? I'll remove from here for now.

Author: Just leaving removed.

=== Upgrade the Beat specification

Any setting can be changed in the Beat yaml specification, including version upgrades. ECK detects those changes and ensures a smooth rolling upgrade of all Beat Pods. Depending on specification settings used, ECK will set the <<{p}-beat-set-beat-output,output>> part of the config, perform Kibana dashboard <<{p}-beat-set-up-kibana-dashboards,setup>>, restart Beats on certificates rollover and set up Beats <<{p}-beat-secrets-keystore-for-secure-settings,keystore>>.

[id="{p}-beat-custom-configuration"]
=== Customize Beat configuration

The Beat configuration is provided through the `config` element:

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: heartbeat-quickstart
spec:
  type: heartbeat
  version: {version}
  elasticsearchRef:
    name: elasticsearch-quickstart
  config:
    heartbeat.monitors:
    - type: tcp
      schedule: '@every 5s'
      hosts: ["elasticsearch-quickstart-es-http.default.svc:9200"]
  deployment:
    podTemplate:
      spec:
        dnsPolicy: ClusterFirstWithHostNet
        securityContext:
          runAsUser: 0
----

> Alternatively, it can be provided indirectly via a Secret specified in the `configRef` element:

Reviewer 3: I would remove "indirectly" which is a bit confusing

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: heartbeat-quickstart
spec:
  type: heartbeat
  version: 7.8.0
  elasticsearchRef:
    name: elasticsearch-quickstart
  configRef:
    secretName: heartbeat-config
  deployment:
    podTemplate:
      spec:
        dnsPolicy: ClusterFirstWithHostNet
        securityContext:
          runAsUser: 0
---
apiVersion: v1
kind: Secret
metadata:
  name: heartbeat-config
stringData:
  beat.yml: |-
    heartbeat.monitors:
    - type: tcp
      schedule: '@every 5s'
      hosts: ["elasticsearch-quickstart-es-http.default.svc:9200"]
----

For more details about Beats configuration, see the link:https://www.elastic.co/guide/en/beats/libbeat/current/beats-reference.html[Beats documentation].

[id="{p}-beat-deploy-elastic-beat"]

> === Deploy Elastic Beat

Author: I took it from ["Elastic provides Beats for capturing..."](https://www.elastic.co/guide/en/beats/libbeat/current/beats-reference.html) - if we provide them, they're ours? But yeah, not an offical name. Maybe just "Deploy a Beat" vs "Deploy a community Beat"?

ECK supports deploying the following Elastic Beats:

- Filebeat
- Metricbeat
- Heartbeat
- Journalbeat
- Auditbeat
- Packetbeat

> The Beat to deploy and its version has to be specified via `type` and `version` specification elements respectively. ECK will compute the correct path to the container image, so `image` element doesn't have to be provided. ECK will create a new user in Elasticsearch with a minimal set of appropriate roles and permissions that allow to use all Beat features.

Reviewer 4: ```suggestion
You can specify the Beat to deploy and its version through `type` and `version` elements. ECK computes the correct path to the container image, so the `image` element doesn't have to be provided. ECK creates a new user in Elasticsearch with a minimal set of appropriate roles and permissions to enable the use all Beat features.
```

[id="{p}-beat-deploy-community-beat"]

> === Deploy community Beat

Author: The capitalization is not consistent in our docs, but I'm fine with C.

> ECK supports the deployment of any community Beat. As with Elastic Beats, `type` and `version` specification elements have to be provided. In addition:

Reviewer: ```suggestion
ECK supports the deployment of any Community Beat. As with Elastic Beats, `type` and `version` specification elements have to be provided. In addition:
```
maybe check with the Beats doc writers regarding "Elastic Beats"

Author: Will do. Also note, it's not all of them. We are not supporting functionbeat and winlogbeat.

> ECK supports deploying any community Beat. As with Elastic Beats, `type` and `version` specification elements have to be provided. In addition:

Reviewer 4: ```suggestion
ECK supports the deployment of any community Beat. As with Elastic Beats, `type` and `version` specification elements have to be provided. In addition:
```

- `image` must point to the image to be deployed

> -  role must exist and have the permissions required by the Beat. The role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.

Reviewer: Reading this it is not clear to me what your refer to with `role`. Maybe some more explanation?

Author: ```suggestion
-  a role must exist in Elasticsearch and have the permissions required by the Beat. The role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.
```
Maybe?

> - `image` has to point to the image to be deployed, and
> - a role has to exist and have the permissions required by the Beat. Role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat` create role named `eck_beat_kafkabeat_role`.

Reviewer 4: ```suggestion
- `image` must point to the image to be deployed
-  role must exist and have the permissions required by the Beat. The role name has to be `eck_beat_type_role`, where `type` is the Beat type. For example, when deploying `type: kafkabeat`, the role name would be `eck_beat_kafkabeat_role`.
```

Author: Not related to the wording, but I'll just mention for everyone here that this assumes [this](https://github.com/elastic/cloud-on-k8s/issues/3363) issue is fixed.

[id="{p}-beat-set-up-kibana-dashboards"]
=== Set up Kibana dashboards

> ECK can instruct Beats to set up an example dashboards packaged with the Beat. To enable this, set `kibanaRef` element in the specification to point to ECK-managed Kibana deployment:

Reviewer: ```suggestion
ECK can instruct Beats to set up example dashboards packaged with the Beat. To enable this, set the `kibanaRef` element in the specification to point to ECK-managed Kibana deployment:
```

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: filebeat-quickstart
spec:
  kibanaRef:
    name: kibana-quickstart
...
----

> ECK will create a new user in Elasticsearch with minimal set of appropriate roles and permissions that is needed for dashboard setup.

Reviewer: ```suggestion
ECK will create a new user in Elasticsearch with a minimal set of appropriate roles and permissions that is needed for dashboard setup.
```

[id="{p}-beat-secrets-keystore-for-secure-settings"]
=== Secrets keystore for secure settings

> Beats offer a secret keystore that can be used when sensitive settings, such as passwords, need to be provided in the configuration. This avoids storing them in the config directly.

Reviewer 4: ```suggestion
Beats offer a secret keystore for sensitive settings that need to be provided in the configuration, like for example passwords. This avoids storing them in the config directly.
```

> ECK exposes that mechanism with `secureSettings` element in the specification. <<{p}-es-secure-settings,Similar to Elasticsearch>>, you can use Kubernetes Secrets to manage it:

Reviewer: ```suggestion
ECK exposes that mechanism with the `secureSettings` element in the specification. <<{p}-es-secure-settings,Similar to Elasticsearch>>, you can use Kubernetes Secrets to provide the settings in a secure way:
```

Not sure about my suggestion maybe @alaudazzi can help? I found the original word `manage` to unspecific.

Reviewer 4: LGTM

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: filebeat-quickstart
spec:
  secureSettings:
  - secretName: agent-name-secret
  config:
    name: ${AGENT_NAME_VAR}
...
---
apiVersion: v1
kind: Secret
metadata:
  name: agent-name-secret
stringData:
  AGENT_NAME_VAR: id_007
----

See Beats link:https://www.elastic.co/guide/en/beats/filebeat/current/keystore.html[documentation] for more details.

[id="{p}-beat-set-beat-output"]
=== Set Beat output

> If `elasticsearchRef` element is specified, ECK populates the output section of the Beat config. ECK creates a user with appropriate roles and permissions and use its credentials. If required, it also mounts the CA certificate in all Beat Pods, and roll over Pods when this certificate changes.

Reviewer: ```suggestion
If the `elasticsearchRef` element is specified, ECK populates the output section of the Beat config. ECK creates a user with appropriate roles and permissions and uses its credentials. If required, it also mounts the CA certificate in all Beat Pods, and recreates Pods when this certificate changes.
```

> If `elasticsearchRef` element is specified ECK will populate output section of the Beat config. ECK will create a user with appropriate roles and permissions and use its credentials. It will also mount CA certificate in all Beat Pods if necessary and roll over Pods when this certificate changes.

Reviewer 4: ```suggestion
If `elasticsearchRef` element is specified, ECK populates the output section of the Beat config. ECK creates a user with appropriate roles and permissions and use its credentials. If required, it also mounts the CA certificate in all Beat Pods, and roll over Pods when this certificate changes.
```

> Output can be set to any that's supported by given Beat. To do that, remove `elasticsearchRef` element from the specification and include appropriate output configuration in the `config` or `configRef`.

Reviewer: ```suggestion
Output can be set to any that's supported by a given Beat. To do that, remove the `elasticsearchRef` element from the specification and include an appropriate output configuration in the `config` or `configRef`.
```

[source,yaml,subs="attributes,+macros"]
----
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: filebeat-quickstart
spec:
  config:
    output.redis:
      hosts: ["my_redis.default.svc"]
      password: "my_password"
      key: "filebeat"
      db: 0
      timeout: 5
...
----

[id="{p}-beat-chose-the-deployment-model"]
=== Choose the deployment model

> Depending on the use case, Beats may need to be deployed as a Deployment or a DaemonSet. Provide `podTemplate` element under either `deployment` or `daemonSet` element in the specification to choose how a given Beat should be deployed.

Reviewer: ```suggestion
Depending on the use case, Beats may need to be deployed as a Deployment or a DaemonSet. Provide a `podTemplate` element under either the `deployment` or the `daemonSet` element in the specification to choose how a given Beat should be deployed.
```

  config:
    filebeat:
      autodiscover:
        providers:
        - host: ${HOSTNAME}
          type: kubernetes
          hints:
            enabled: true
            default_config:
              type: container
              paths:
              - /var/log/containers/*${data.kubernetes.container.id}.log
  daemonSet:
    podTemplate:
      spec:
        serviceAccount: elastic-beat-filebeat-quickstart
        automountServiceAccountToken: true
...
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elastic-beat-filebeat-quickstart
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: elastic-beat-autodiscover-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: elastic-beat-autodiscover
subjects:
- kind: ServiceAccount
  name: elastic-beat-filebeat-quickstart
  namespace: default
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: elastic-beat-autodiscover
rules:
- apiGroups:
  - ""
  resources:
  - nodes
  - namespaces
  - events
  - pods
  verbs:
  - get
  - list
  - watch
----

[id="{p}-beat-deploying-beats-in-secured-clusters"]
=== Deploying Beats in secured clusters

> To deploy Beats in clusters with Pod Security Policy admission controller enabled, or in OpenShift clusters, you have to grant additional permissions to the Service Account that Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.

Reviewer: ```suggestion
To deploy Beats in clusters with the Pod Security Policy admission controller enabled, or in OpenShift clusters, you have to grant additional permissions to the Service Account that the Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.
```

> Deploying Beats in clusters with Pod Security Policy admission controller enabled or in OpenShift clusters require granting additional permissions to the Service Account that Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.

Reviewer 4: ```suggestion
To deploy Beats in clusters with Pod Security Policy admission controller enabled, or in OpenShift clusters, you have to grant additional permissions to the Service Account that Beat Pods are running with. Those Service Accounts need to be bound to a Role or ClusterRole that has `use` permission for the required PSP/SCC. Different Beats and their features might require different settings set in their PSP/SCC.
```

[id="{p}-beat-troubleshooting"]
== Troubleshooting

[id="{p}-beat-beat-pods-are-crashing-when-kibanaref-is-specified"]
=== Beat Pods are crashing when kibanaRef is specified

> When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Pod crashes. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The pod may cycle at most few times during the first initialization. Afterwards, the Beat should run successfully.

Reviewer: Is crash the right word? It exits in a controlled manner with an error code, not sure what the right word is though.

Author: ```suggestion
When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Beat process exits and the Pod restarts. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The pod may cycle at most few times during the first initialization. Afterwards, the Beat should run successfully.
```
Maybe?

> When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Pod crashes. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The pod may cycle at most few times during the first initialization. Afterwards, the Beat should run successfully.

Reviewer: ```suggestion
When `kibanaRef` is specified, Beat tries to connect to the Kibana instance. If it's unable to do so, the Pod crashes. This may happen when Kibana is not yet up or when Beat user is not yet created in Elasticsearch. The pod may cycle at most a few times during the first initialization. Afterwards, the Beat should run successfully.
```

[id="{p}-beat-configuration-containing-key-null-is-malformed"]
=== Configuration containing key: null is malformed

> When Beat config containing `some_key: null` is applied using `kubectl apply` it's interpreted by `kubectl` diff mechanism as an instruction to remove key `some_key` from the config. This is unfortunate, but works according to the spec it follows. To overcome that, `null` can sometimes be replaced by an empty dictionary, like: `some_key: {}`. If this is not possible due to a particular settings semantics, `configRef` field can be used instead of `config` as Secret data doesn't go through that diff process.

Author: @alaudazzi How about this?

```suggestion
When `kubectl` is used to modify a resource, it calculates a diff between the user applied and the already existing config. This diff has a special semantics link:https://tools.ietf.org/html/rfc7396#section-1[defined], that allow forcing removal of a particular key by setting its value to `null`. When user applied config contains `some_key: null` (or equivalent `some_key: ~`) it's interpreted as an instruction to remove `some_key`, often resulting in a malformed config. In Beat configs, this is often the case with link:https://www.elastic.co/guide/en/beats/filebeat/current/add-cloud-metadata.html[processors]. To avoid this:

- change `some_key: null` to `some_key: {}` - if it doesn't change config semantics
- instead of `config` use `configRef` and reference a Secret containing the config - Secret data doesn't go through the same diffing process
```

Reviewer 4: With more details, it's definitely better.
For 
`that allow forcing removal of a particular key by setting its value to `null``
maybe we can reword into
`that allows you to force the removal of a particular key by setting its value to `null`
?

> When Beat config containing `some_key: null` is applied using `kubectl apply` it's interpretted by `kubectl` diffing mechanism as an instruction to remove key `some_key` from the config. This is unfortunate, but works according to the spec it follows. To overcome that, `null` can sometimes be replaced by an empty dictionary, like: `some_key: {}`. If this is not possible due to a particular settings semantics, `configRef` field can be used instead of `config` as Secret data doesn't go through that diffing process.

Reviewer 4: ```suggestion
When Beat config containing `some_key: null` is applied using `kubectl apply` it's interpreted by `kubectl` diff mechanism as an instruction to remove key `some_key` from the config. This is unfortunate, but works according to the spec it follows. To overcome that, `null` can sometimes be replaced by an empty dictionary, like: `some_key: {}`. If this is not possible due to a particular settings semantics, `configRef` field can be used instead of `config` as Secret data doesn't go through that diff process.
```
Except for a couple of typos (interpretted => interpreted), the first two lines are a bit unclear. We might try to reword them.

[id="{p}-beat-secrets-keystore-for-secure-settings"]
=== Secrets keystore for secure settings

> Beats offer a secret keystore for sensitive settings that need to be provided in the config, like for example passwords. This avoids storing them in the config directly.

Reviewer 2: ```suggestion
Beats offer a secret keystore for sensitive settings that need to be provided in the config, for example passwords. This avoids storing them in the config directly.
```
